### PR TITLE
[SPARK-40646][SQL] Fix returning partial results in JSON data source and JSON functions

### PR DESCRIPTION
**Pull request by sadikovi**
### What changes were proposed in this pull request?

This PR is a follow-up for [SPARK-33134](https://issues.apache.org/jira/browse/SPARK-33134) (https://github.com/apache/spark/pull/30031). 

I found another case when, depending on the order of columns, parsing one JSON field breaks all of the subsequent fields resulting in all nulls:

With a file like this:
```
{"a": {"x": 1, "y": true}, "b": {"x": 1}}
{"a": {"x": 2}, "b": {"x": 2}}
```

Reading the file results in column `b` as null even though it is a valid column.
```scala
val df = spark.read
  .schema("a struct<x: int, y: struct<x: int>>, b struct<x: int>")
  .json("path") 

===

a	                b
null	                null
{"x":2,"y":null}	{"x":2} 
```

However, b column should be:
```
{"x": 1}
{"x": 2}
```

This particular example actually used to work in earlier Spark versions but it was affected by SPARK-33134 which fixed another bug with the incorrect parsing in `from_json`. Because this case was not tested, we missed it at the time. 

In order to fix both SPARK-33134 and SPARK-40646, we need to process `PartialResultException` in `convertArray` method to handle any errors in child objects. Without the fix, the code would not wrap the row in the array for `from_json` resulting in a ClassCastException (SPARK-33134). Because of this handling, we don't need `isRoot` check anymore in `convertObject` thus unblocking SPARK-40646.

I updated the code to handle both cases. With these changes, we can correctly parse this case:
```scala
val df3 = Seq("""[{"c2": [19], "c1": 123456}]""").toDF("c0")
checkAnswer(df3.select(from_json($"c0", ArrayType(st))), Row(Array(Row(123456, null))))
```
which was previously returning `null` for the root row.

### Why are the changes needed?

Fixes a long-standing issue when parsing a JSON with an incorrect field that would break parsing of the entire record.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

I added unit tests for SPARK-40646 as well as SPARK-33134.

### CODE REVIEW

**Comment by sadikovi** [comment]
cc @MaxGekk @HyukjinKwon, I would appreciate a review. Let me know if you have any concerns with the changes. Thank you.

**Comment by MaxGekk** [comment]
My related PR https://github.com/apache/spark/pull/23253

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 1]
```diff
@@ -853,11 +853,68 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
...
     val df4 = Seq("""{"c2": [19]}""").toDF("c0")
     checkAnswer(df4.select(from_json($"c0", MapType(StringType, st))), Row(null))
   }
 
+  test("SPARK-40646: return partial results for JSON arrays with objects") {
+    val st = new StructType()
+      .add("c1", StringType)
+      .add("c2", ArrayType(new StructType().add("a", LongType)))
+
+    // Value of "c2.a" is a string instead of a long.
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:866]: Could you explain this comment `"c2.a" is a string`, and doesn't it contradict to `"c2": {"a": 1}`. Is `1` string?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala [hunk 2]
```diff
@@ -456,7 +456,7 @@ class JacksonParser(
             schema.existenceDefaultsBitmask(index) = false
           } catch {
             case e: SparkUpgradeException => throw e
-            case NonFatal(e) if isRoot =>
+            case NonFatal(e) =>
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:459]: If you allow this, please, check all complex types `map`, `struct`. Or have you added some tests already?

Comment (sadikovi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:459]: I have tried covering different cases in:
- SPARK-40646: return partial results for JSON arrays with objects
- SPARK-40646: return partial results for objects with values as JSON arrays

I think I am missing a test for map, will add, thanks.

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala (hunk 1 above)
Comment (sadikovi) [reply sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:866]: Thanks for pointing it out, the comment is incorrect, it belonged to the original test that I refactored later. I will update.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala (hunk 2 above)
Comment (sadikovi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:459]: I added a test!

**Comment by sadikovi** [comment]
@MaxGekk I have addressed your comments, could you review again?

I think there is still an issue with complex fields:
```scala
test("SPARK-40646") {
  val st = new StructType()
    .add("c1", MapType(StringType, new StructType().add("a1", LongType)))
    .add("c2", StringType)

  // Map "c2" has "k2" key that is a string, not an integer.
  val df = Seq("""{"c1": {"k1": {"a1": 1}, "k2": {"a1": "A"}, "k3": {"a1": 2}}}, "c2": "abc"}""").toDF("c0")
  checkAnswer(
    df.select(from_json($"c0", st)),
    Row(Row(null, "abc"))
  )
}
```

would return `null, null` instead of `null, "abc"`. 

Would it be okay to address that case in the follow-up? It might require a bit of time to fix that case.

**Comment by sadikovi** [comment]
cc @cloud-fan

**Comment by MaxGekk** [comment]
> Would it be okay to address that case in the follow-up?

Sure.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @sadikovi.

**Comment by sadikovi** [comment]
Thanks for merging!

**Pull request closed without merging**